### PR TITLE
Improve controller

### PR DESCRIPTION
**Pull request by lynxthecat**
This pull request is intended to capture improvements to the controller including detection of bufferbloat and adjustments to the shaper rates. Addresses https://github.com/lynxthecat/cake-autorate/issues/225.

### CODE REVIEW

**Comment by lynxthecat** [comment]
@moeller0 how about this as a first step? Since your proposal involves working with averages, it makes sense to me to switch to working with the average delay across the bufferbloat detection window. Can you agree with this first step before I introduce the proportional reductions as you outlined here: https://github.com/lynxthecat/cake-autorate/commit/6a8c740b8ec53d14b5db0b807ab53744ee31887e#commitcomment-122905855?

**Comment by lynxthecat** [comment]
But @moeller0 we will rotate out any bad reflectors already with our rotator health checking code.

**Comment by moeller0** [comment]
On 28 July 2023 13:34:32 CEST, lynxthecat ***@***.***> wrote:
>But @moeller0 we will rotate out any bad reflectors already with our rotator health checking code.
>


This operates on a different timescale... as far as I am concerned the consensus over a diverse set of reflectors is one of the core ideas of autorate that proved its merit, I would be very careful in changing that and certainly not based on cursory testing alone. Such testing catches really bad ideas quickly but passing such a test does not tell much about the performance in edge cases, but it is these edge cases that affect how well this thing be'haves under unfortunate loads.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by lynxthecat** [comment]
What tests would you recommend I try?

**Comment by moeller0** [comment]
Gesendet: Freitag, 28. Juli 2023 um 15:50 Uhr
Von: "lynxthecat" ***@***.***>
An: "lynxthecat/cake-autorate" ***@***.***>
Cc: "moeller0" ***@***.***>, "Mention" ***@***.***>
Betreff: Re: [lynxthecat/cake-autorate] Improve controller (PR #248)


 

What tests would you recommend I try?

For subtle changes like replacing the minum over a set of reflectors with average over a set of reflectors I think you would need to run the new code for at leasat a fortnight. But here for me the question is easy to answer, using the minumum over a set makes theoretucal sense, taking the average not.
Question: since we are actually discussing different changes why is this additoinal change so important fir you right now?

**Comment by lynxthecat** [comment]
@moeller0 OK how about this now? I think this is pretty much exactly your idea?

**Comment by lynxthecat** [comment]
@moeller0 OK how about this now? I think this is pretty much exactly your idea?

**Comment by lynxthecat** [comment]
@moeller0 you have convinced me. 

So how about this now (please check the commit(s) / files changed in this pull request)? 

I think this is pretty much exactly your idea.

Now:

- we still use the trusty old bufferbloat detection mechanism 
- we no longer consult the achieved rate when setting the shaper rate - this concept has been jettisoned
- the default shaper rate adjustment has been tweaked from 0.9 to 0.85
- we additionally keep track of average delta owd across the detection window
- on bufferbloat we scale the the shaper rate adjustment based upon average delta owd / max average delta owd  
- thus if the average delta owd >= max average delta owd, the full 0.85 adjustment is employed
- if the average delta owd us < max average delta owd, an adjustment between 0.85 and 1.0 is employed in proportion to the ratio between the average delta owd and the max average delta owd

Here are some example values from a speed test:

https://gist.github.com/lynxthecat/b92810b6cc1687d820c2cc5e380e8c4e

**Comment by lynxthecat** [comment]
@moeller0 you have convinced me.

So how about this now (please check the commit(s) / files changed in this pull request)?

I think this is pretty much exactly your idea.

Now:

we still use the trusty old bufferbloat detection mechanism
we no longer consult the achieved rate when setting the shaper rate - this concept has been jettisoned
we additionally keep track of average delta owd across the detection window
on bufferbloat we scale the the shaper rate adjustment based upon average delta owd / max average delta owd
thus if the average delta owd >= max average delta owd, the full 0.90 adjustment is employed
if the average delta owd us < max average delta owd, an adjustment between 0.90 and 1.0 is employed in proportion to the ratio between the average delta owd and the max average delta owd
Here are some example values from a speed test:

https://gist.github.com/lynxthecat/b92810b6cc1687d820c2cc5e380e8c4e

**Comment by lynxthecat** [comment]
@moeller0 you have convinced me.

So how about this now (please check the commit(s) / files changed in this pull request)?

I think this is pretty much exactly your idea.

Now:

we still use the trusty old bufferbloat detection mechanism
we no longer consult the achieved rate when setting the shaper rate - this concept has been jettisoned
we additionally keep track of average delta owd across the detection window
on bufferbloat we scale the the shaper rate adjustment based upon average delta owd / max average delta owd
thus if the average delta owd >= max average delta owd, the full 0.90 adjustment is employed
if the average delta owd us < max average delta owd, an adjustment between 0.90 and 1.0 is employed in proportion to the ratio between the average delta owd and the max average delta owd
Here are some example values from a speed test:

https://gist.github.com/lynxthecat/b92810b6cc1687d820c2cc5e380e8c4e

**Comment by lynxthecat** [comment]
@moeller0 you have convinced me.

So how about this now (please check the commit(s) / files changed in this pull request)?

I think this is pretty much exactly your idea.

Now:

we still use the trusty old bufferbloat detection mechanism
we no longer consult the achieved rate when setting the shaper rate - this concept has been jettisoned
we additionally keep track of average delta owd across the detection window
on bufferbloat we scale the the shaper rate adjustment based upon average delta owd / max average delta owd
thus if the average delta owd >= max average delta owd, the full 0.90 adjustment is employed
if the average delta owd us < max average delta owd, an adjustment between 0.90 and 1.0 is employed in proportion to the ratio between the average delta owd and the max average delta owd
Here are some example values from a speed test:

https://gist.github.com/lynxthecat/b92810b6cc1687d820c2cc5e380e8c4e

**Comment by lynxthecat** [comment]
@moeller0 you have convinced me.

So how about this now (please check the commit(s) / files changed in this pull request)?

I think this is pretty much exactly your idea.

Now:

we still use the trusty old bufferbloat detection mechanism
we no longer consult the achieved rate when setting the shaper rate - this concept has been jettisoned
we additionally keep track of average delta owd across the detection window
on bufferbloat we scale the the shaper rate adjustment based upon average delta owd / max average delta owd
thus if the average delta owd >= max average delta owd, the full 0.90 adjustment is employed
if the average delta owd us < max average delta owd, an adjustment between 0.90 and 1.0 is employed in proportion to the ratio between the average delta owd and the max average delta owd
Here are some example values from a speed test:

https://gist.github.com/lynxthecat/b92810b6cc1687d820c2cc5e380e8c4e

**Comment by moeller0** [comment]
> I think this is pretty much exactly your idea.

Hardly, this is just a dust-off of a proposal you made some time ago, I take responsibility for helping define the initial algorithm, but this is IMHO a cooperative idea.

**Comment by lynxthecat** [comment]
What do you think of these defaults:

```
# owd delta threshold in ms is the extent of OWD increase to classify as a delay
# these are automatically adjusted based on maximum on the wire packet size
# (adjustment significant at sub 12Mbit/s rates, else negligible)
dl_owd_delta_thr_ms=30.0 # (milliseconds)
ul_owd_delta_thr_ms=30.0 # (milliseconds)

# average owd delta threshold in ms at which maximum adjust_down_bufferbloat is applied
dl_avg_owd_delta_thr_ms=60.0 # (milliseconds)
ul_avg_owd_delta_thr_ms=60.0 # (milliseconds)
```

**Comment by moeller0** [comment]
These likely need to be adapted for each link. It seems OK if you distribute values that work reasonably on your link and instruct users at seriously considering changing those...
That said, I have a feel for how to set the xx_owd_delta_thr_ms values by looking at CDFs with and without load, but I have nointuition yet for the xx_avg_owd_delta_thr_ms values... my gut feeling is 60ms might be a bit high but realistically one needs to look at the actual max adjustment factor as well... This just needs some testing...

File: cake-autorate.sh [hunk 1]
```diff
@@ -374,9 +374,18 @@ update_shaper_rate()
 		*bb*)
```
Comment (lynxthecat) [inline cake-autorate.sh:374]: @moeller0 can you see a way to simplify the logic and/or calculations below?

**Comment by lynxthecat** [comment]
So far so good @moeller0. This change in terms of bufferbloat reduction scaling seems to be working very well. I'm tempted to call it a day with these changes and pull them in to master. Then after a week or two more we could issue a v3.1.0 release with these new changes.

**Comment by moeller0** [comment]
> So far so good @moeller0. This change in terms of bufferbloat reduction scaling seems to be working very well.

So your intuition a while ago seems fully validated.

> I'm tempted to call it a day with these changes and pull them in to master. Then after a week or two more we could issue a v3.1.0 release with these new changes.

As I said, you should expose them to enough testing to be sure you covered not only the simple test cases (like speed-tests), but after that the best way forward seems to broaden the testing base and if that means pushing to "mater" (still like this better than the alternatives) then so be it.

**Comment by lynxthecat** [comment]
@bairhys this is very likely to get pulled into master unless there emerge specific issues with it. Would you mind switching over to it:

```
setup.sh lynxthecat/cake-autorate improve-controller 
```

If you previously specified a custom `dl_delay_thr_ms` or `ul_delay_thr_ms` those have changed to `dl_owd_delta_thr_ms` and `ul_owd_delta_thr_ms`, and you may then want to specify values for `dl_avg_owd_delta_thr_ms` and `ul_avg_owd_delta_thr_ms` that are around twice whatever you set for the owd_delta_thr_ms values. Also the DATA log lines have been tweaked slightly (check out the new headers).

**Comment by bairhys** [comment]
Hi @lynxthecat, I am running this branch now. I made a [new branch](https://github.com/bairhys/prometheus-cake-autorate-exporter/tree/Improve-controller) for my exporter to read the updated DATA and SUMMARY log lines. The rate of development of this project is really impressive!

**Comment by lynxthecat** [comment]
Thanks @bairhys any comment on performance? In any case, I'd love to see some plots with the new elements. Could you possibly post here?

**Comment by bairhys** [comment]
I did some back to back testing comparing v3.0.1 to this branch `improve-controller` and later no sqm.

Testing setup was 

- `improve-controller` branch in /root/cake-autorate
- I downloaded `3.0.1` release, put this at /root/cake-autorate-v.3.0.1, 
- changed `cake-autorate.sh` to look at folder `PREFIX=/root/cake-autorate-3.0.1`
- both settings were same, see end of post for settings for both
- 2 ssh sessions, left session to run `v3.0.1`, right to run `improve-controller`
- 2 chrome windows, left for `v3.0.1` tests, right for `improve-controller` tests
- after each test, switched cake-autorate version

Results

All results summarized here with a score given to all tests.

I defined one as overall score `download speed` * `upload speed` / (`download lat` * `upload lat`)
Then another as a throughput score `download speed` * `upload speed` 
And last as latency score `download lat` * `upload lat`

| V3.0.1             |      |        |        |                                              |                                |                                       |
|--------------------|------|--------|--------|----------------------------------------------|--------------------------------|---------------------------------------|
| dl                 | up   | dl lat | ul lat | Rating 1 (dl*ul/dl_lat/ul_lat) higher better | Rating 2 (dl*ul) higher better | Rating 3 (dl_lat*ul_lat) lower better |
| 66.7               | 5.64 | 17     | 14     | 1.5806218487395                              | 376.188                        | 238                                   |
| 52.3               | 5.69 | 14     | 15     | 1.41708095238095                             | 297.587                        | 210                                   |
| 65.8               | 8.25 | 18     | 14     | 2.15416666666667                             | 542.85                         | 252                                   |
| 66.5               | 8.23 | 16     | 12     | 2.85049479166667                             | 547.295                        | 192                                   |
|                    |      |        | TOTAL  | 8.00236425945378                             | 1763.92                        | 892                                   |
| improve-controller |      |        |        |                                              |                                |                                       |
| 73                 | 5.7  | 19     | 13     | 1.68461538461538                             | 416.1                          | 247                                   |
| 70.1               | 7.84 | 18     | 13     | 2.34864957264957                             | 549.584                        | 234                                   |
| 68.6               | 8.25 | 16     | 14     | 2.5265625                                    | 565.95                         | 224                                   |
| 66.7               | 8.56 | 17     | 14     | 2.39895798319328                             | 570.952                        | 238                                   |
|                    |      |        | TOTAL  | 8.95878544045823                             | 2102.586                       | 943                                   |
| No-sqm             |      |        |        |                                              |                                |                                       |
| dl                 | up   | dl lat | ul lat |                                              |                                |                                       |
| 79.1               | 8.31 | 62     | 234    | 0.0453074855252275                           | 657.321                        | 14508                                 |

The extra latency between the `improve-controller` and `v3.0.1` tests are similar even though `v3.0.1` is a bit better.

`improve-controller` had slightly better throughput initially then they started to converge.

Overall the difference in additional latency is minimal so based on throughput `improve-controller` looks to have a slight edge but need more tests to confirm. I could see this test becoming automated to give better and less time consuming long term test results.

raw data

test 1 `v3.0.1`
image

test 2 `improve-controller` 
image

test 3 `v3.0.1`
image

test 4  `improve-controller` 
image

test 5 `v3.0.1`
image

test 6 `improve-controller` 
image

test 7 `v3.0.1`
image

test 8 `improve-controller` 
image

test 9 `no-sqm`
image






```
#!/bin/bash

# *** STANDARD CONFIGURATION OPTIONS ***

output_processing_stats=1 	# enable (1) or disable (0) output monitoring lines showing processing stats
output_load_stats=1       	# enable (1) or disable (0) output monitoring lines showing achieved loads
output_reflector_stats=1  	# enable (1) or disable (0) output monitoring lines showing reflector stats
output_cake_changes=1     	# enable (1) or disable (0) output monitoring lines showing cake bandwidth changes
debug=1 		  	# enable (1) or disable (0) out of debug lines


### For multihomed setups, it is the responsibility of the user to ensure that the probes
### sent by this instance of cake-autorate actually travel through these interfaces.
### See ping_extra_args and ping_prefix_string

dl_if=ifb4wan  # download interface
ul_if=wan     # upload interface

# Set either of the below to 0 to adjust one direction only
# or alternatively set both to 0 to simply use cake-autorate to monitor a connection
adjust_dl_shaper_rate=1 # enable (1) or disable (0) actually changing the dl shaper rate
adjust_ul_shaper_rate=1 # enable (1) or disable (0) actually changing the ul shaper rate

min_dl_shaper_rate_kbps=30000  # minimum bandwidth for download (Kbit/s)
base_dl_shaper_rate_kbps=50000 # steady state bandwidth for download (Kbit/s)
max_dl_shaper_rate_kbps=100000  # maximum bandwidth for download (Kbit/s)

min_ul_shaper_rate_kbps=4000  # minimum bandwidth for upload (Kbit/s)
base_ul_shaper_rate_kbps=6000 # steady state bandwidth for upload (KBit/s)
max_ul_shaper_rate_kbps=15000  # maximum bandwidth for upload (Kbit/s)

# *** OVERRIDES ***


randomize_reflectors=0

pinger_binary=tsping
```

**Comment by bairhys** [comment]
Here are some plots, the new DATA columns are near the bottom 

test

**Comment by lynxthecat** [comment]
All looks great to me. You could play around with the defaults such as by lowering the average owd delta threshold if you wanted to reduce latency (and the old behaviour can be obtained by just setting those values to 0). But the defaults look good to me.
